// msa/lib/model/FeatureCol.js
// Generated by CoffeeScript 1.9.0
var Collection, Feature, FeatureCol, _;

Feature = require("./Feature");

Collection = require("backbone-thin").Collection;

_ = require("underscore");

module.exports = FeatureCol = Collection.extend({
  model: Feature,
  constructor: function() {
    this.startOnCache = [];
    this.on("all", function() {
      return this.startOnCache = [];
    }, this);
    return Collection.apply(this, arguments);
  },
  startOn: function(index) {
    if (this.startOnCache[index] == null) {
      this.startOnCache[index] = this.where({
        xStart: index
      });
    }
    return this.startOnCache[index];
  },
  contains: function(index) {
    return this.reduce(function(el, memo) {
      return memo || el.contains(index);
    }, false);
  },
  getFeatureOnRow: function(row, x) {
    return this.filter(function(el) {
      return el.get("row") === row && el.get("xStart") <= x && x <= el.get("xEnd");
    });
  },
  assignRows: function() {
    var len, rows, x;
    len = (this.max(function(el) {
      return el.get("xEnd");
    })).attributes.xEnd;
    rows = (function() {
      var _i, _results;
      _results = [];
      for (x = _i = 0; 0 <= len ? _i <= len : _i >= len; x = 0 <= len ? ++_i : --_i) {
        _results.push(0);
      }
      return _results;
    })();
    this.each(function(el) {
      var max, _i, _ref, _ref1;
      max = 0;
      for (x = _i = _ref = el.get("xStart"), _ref1 = el.get("xEnd"); _i <= _ref1; x = _i += 1) {
        if (rows[x] > max) {
          max = rows[x];
        }
        rows[x]++;
      }
      return el.set("row", max);
    });
    return _.max(rows);
  },
  getCurrentHeight: function() {
    return (this.max(function(el) {
      return el.get("row");
    })).attributes.row + 1;
  },
  getMinRows: function() {
    var len, rows, x;
    len = (this.max(function(el) {
      return el.get("xEnd");
    })).attributes.xEnd;
    rows = (function() {
      var _i, _results;
      _results = [];
      for (x = _i = 0; 0 <= len ? _i <= len : _i >= len; x = 0 <= len ? ++_i : --_i) {
        _results.push(0);
      }
      return _results;
    })();
    this.each(function(el) {
      var _i, _ref, _ref1, _results;
      _results = [];
      for (x = _i = _ref = el.get("xStart"), _ref1 = el.get("xEnd"); _i <= _ref1; x = _i += 1) {
        _results.push(rows[x]++);
      }
      return _results;
    });
    return _.max(rows);
  }
});
